src/mesh/refinement_cc_in_one.hpp

```diff
@@ -29,13 +30,186 @@
+inline void ProlongationRestrictionLoop(const cell_centered_bvars::BufferCacheHost_t &info_h,
+                                        const IndexShape &cellbounds,
+                                        const IndexShape &c_cellbounds,
+                                        const RefinementOp_t op) {
+  const IndexDomain interior = IndexDomain::interior;
+  auto ckb = c_cellbounds.GetBoundsK(interior);
+  auto cjb = c_cellbounds.GetBoundsJ(interior);
+  auto cib = c_cellbounds.GetBoundsI(interior);
+  auto kb = cellbounds.GetBoundsK(interior);
+  auto jb = cellbounds.GetBoundsJ(interior);
+  auto ib = cellbounds.GetBoundsI(interior);
+  const int nbuffers = info_h.extent_int(0);
+  for (int buf = 0; buf < nbuffers; ++buf) {
+      if (DoRefinementOp(info_h(buf), op)) {
+        int sk, ek, sj, ej, si, ei;
+        GetLoopBoundsFromBndInfo<DIM>(info_h(buf), ckb.s, cjb.s, sk, ek, sj, ej, si, ei);
+        auto coords = info_h(buf).coords;
+        auto coarse_coords = info_h(buf).coarse_coords;
+        auto coarse = info_h(buf).coarse;
+        auto fine = info_h(buf).fine;
+        par_for(
+            DEFAULT_LOOP_PATTERN, "ProlongateOrRestrictCellCenteredValues",
+            DevExecSpace(), 0, info_h(buf).Nt - 1, 0, info_h(buf).Nu - 1, 0,
+            info_h(buf).Nv - 1, sk, ek, sj, ej, si, ei,
+            KOKKOS_LAMBDA(const int l, const int m, const int n, const int k, const int j,
+                          const int i) {
+              Stencil<DIM>::Do(l, m, n, k, j, i, ckb, cjb, cib, kb, jb, ib,
+                               coords, coarse_coords, coarse, fine);
+            });
+      }
+    }
+}
+template <int DIM, template <int> typename Stencil>
+inline void ProlongationRestrictionLoop(const cell_centered_bvars::BufferCache_t &info,
+                                        const cell_centered_bvars::BufferCacheHost_t &info_h,
+                                        const IndexShape &cellbounds,
+                                        const IndexShape &c_cellbounds,
+                                        const RefinementOp_t op) {
+  const int nbuffers = info_h.extent_int(0);
+  if (nbuffers > MIN_NUM_BUFS) {
+    ProlongationRestrictionLoop<DIM, Stencil>(info, cellbounds, c_cellbounds, op);
+  } else {
+    ProlongationRestrictionLoop<DIM, Stencil>(info_h, cellbounds, c_cellbounds, op);
+  }
+}
+
+template<template<int> typename Stencil, typename... Args>
+inline void DoProlongationRestrictionOp(const IndexShape &cellbnds, Args &&...args) {
+  const IndexDomain entire = IndexDomain::entire;
+  if (cellbnds.ncellsk(entire) > 1) { // 3D
+    ProlongationRestrictionLoop<3, Stencil>(std::forward<Args>(args)...);
+  } else if (cellbnds.ncellsj(entire) > 1) { // 2D
+    ProlongationRestrictionLoop<2, Stencil>(std::forward<Args>(args)...);
+  } else if (cellbnds.ncellsi(entire) > 1) { // 1D
+    ProlongationRestrictionLoop<1, Stencil>(std::forward<Args>(args)...);
+  }
+}
+
+} // namespace impl
+
```

Comment:
These are the new unified loop constructs. So you can call, for example,
```C++
DoProlongationRestrictionOp<MyProlongationStencil>(cellbnds, info, info_h, cellbounds, c_cellbounds, RefinementOp_t::Prolongation);
```
and that will automatically do prolongation on the set of buffers/variables, appropriately switching to a coalesced loop or hierarchical parallelism as appropriate, and doing the appropriate dimensionality of the loop.

I know it's a little distasteful to have to include cellbounds twice, but there's a win as far as code size.